### PR TITLE
wasi: Implement more of the standard library

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit fills out more of the `wasm32-unknown-wasi` target's standard library, notably the `std::fs` module and all of its internals. A few tweaks were made along the way to non-`fs` modules, but the last commit contains the bulk of the work which is to wire up all APIs to their equivalent on WASI targets instead of unconditionally returning "unsupported". After this some basic filesystem operations and such should all be working in WASI!

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @fitzgen 
cc @sunfishcode

File: src/libstd/sys/wasi/fs.rs [hunk 1]
```diff
@@ -1,294 +1,689 @@
...
+                    v.len(),
+                ));
+                if res.is_err() {
+                    continue;
+                }
+                let path = PathBuf::from(OsString::from_vec(v));
+                map.insert(path, ManuallyDrop::new(WasiFd::from_raw(fd)));
+            }
+            let ptr = Box::into_raw(map);
+            match PTR.compare_exchange(ptr::null_mut(), ptr, SeqCst, SeqCst) {
```
Comment (fitzgen) [inline src/libstd/sys/wasi/fs.rs]: A comment here explaining how if we lost the race to initialize, we drop our local map would be good. I feel  like I always have to double check for this kind of thing and atomics can be so tricky...

Comment (fitzgen) [reply src/libstd/sys/wasi/fs.rs]: But yes -- getting this shared with upstream would be A+++!

**Review by fitzgen — APPROVED** [review]
LGTM!

**Comment by bluetech** [comment]
I am curious about the preopened map approach.

If I understand correctly, the WASI API (like Capsicum and CloudABI) doesn't support absolute paths, only dirfd-relative paths, but the std::fs API has many absolute-path APIs, but no dirfd-relative APIs. To work around this, this PR maintains a hashmap in the runtime which maps directory path -> dirfd for a set of "preopened files", and then absolute paths are converted to dirfd-relative by prefix matching.

This is pretty clever, and I understand why it is needed for the WASI support to be useful. However, I find it a bit surprising and unusual that Rust should do such things behind the scenes. The "harder" alternative is to design APIs for dirfd-relative operations (wrapping `openat` and friends), implement them in WASI, and not support the absolute-path variants. (Such APIs would be very useful for the other UNIX OSs and CloudABI as well - I really think the capability-friendly APIs are much better for security).

For practical reasons, I think the approach in this PR is a good idea, but maybe this should be discussed later before stabilization.

**Comment by alexcrichton** [comment]
@bluetech good questions! Sounds like you've got a good handle on how things are implemented, so I'll just explain the rationale!

I think that it's not feasible for the wasi target to *not* provide `std::fs::File::open` as it is today. That API (and others) are already widely used throughout the ecosystem and I think we need it to work for better or worse, which necessitates some form of preopend map approach.

Now you're right though in that this isn't true to wasi! Currently this PR adds a suite of extension traits in `std::os::wasi::fs` which enable all the `openat`-style APIs. None of those extensions use the preopened map approach, only the existing APIs use that.

Long-term I think there's a story to be made for stabilizing a cross-platform version of these APIs. For example I think we could add `openat`-style APIs to `std::fs` itself, and implement them with `openat` on Linux and the native syscalls on wasi. That would basically mean that the `std::os::wasi::fs` module would get deprecated in favor of the inherent methods already available in `std::fs`. We're still pretty far away from this though, so I didn't want to try to do that on this PR. 

In the near term it might be interesting to prototype these APIs on crates.io, and envision `std::fs` "as if it only had `openat`". We could try porting crates to that and they'd all work natively without a preopened map on wasi, but it could all be piecemeal and crate by crate.

Hopefully that clears things up!

File: src/libstd/sys/wasi/args.rs [hunk 2]
```diff
@@ -34,18 +19,31 @@ pub struct Args {
 
 /// Returns the command line arguments
 pub fn args() -> Args {
+    maybe_args().unwrap_or_else(|_| {
```
Comment (lachlansneff) [inline src/libstd/sys/wasi/args.rs]: What's the reasoning behind getting the args from the host every time `args()` is called? Wouldn't it make sense to get it once at program startup and store it?

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/59653#issuecomment-479305437, @bors r-

File: src/libstd/sys/wasi/args.rs (hunk 2 above)
Comment (alexcrichton) [reply src/libstd/sys/wasi/args.rs]: Oh sure yeah, the thinking is that this reduces our dependency on a C library somewhat, but it also removes the need for global state and locks. This way if you don't actually use the arguments we never reserve space or allocate memory for them. 

In general I don't think acquiring the arguments isn't too performance critical, so I wanted to make sure that the binary and memory impact would be as small as possible. This also matches what we do on OSX I believe, for example.

**Comment by Centril** [comment]
@bors p=2

Rollup fairness.

**Pull request merged**